minetest/minetest
# Add WordWrapper util (with unit tests, ellipsis truncation, and fixes)

This util is currently used by StaticText and EditBoxes. I plan on making a `label[]` with a size and ellipsis truncation in a future PR - this will be used in ContentDB's dialog.

* Reduce code duplication by creating a word wrap util.
* Fixed support for `\r\n`.
* Add ellipsis (`…`) when text is truncated, to avoid text being cut off.
* Added unit tests.
* Fixed support for soft hyphens (a unicode control character).

Fixes #10285

See also https://en.wikipedia.org/wiki/Line_wrap_and_word_wrap

Long info text:

![image](https://user-images.githubusercontent.com/2122943/105559960-7c377f00-5d0a-11eb-91a1-5277a9ffea5c.png)

## To do

- [x] Fix support for `\r\n`
- [x] Add support for ellipsis
- [x] Fix support for soft hyphens
- [x] Fix not wrapping soon enough in StaticText causing text to clip
- [x] Fix EditBox selection issue
- [x] Readd support for disabling `/n` in singleline text boxes with wrapping
- [ ] DRY-ify WordWrapper
- [ ] ~~Break words~~
- [ ] ~~Add support other unicode control characters~~

## How to test

Run unit tests, try it out with various formspec and infotexts.

See meta editor in https://github.com/minetest/minetest/pull/10431
